### DISCUSSION TITLE
start/stoping AnalogAudioStream and AudioRealFFT

### DISCUSSION THREAD

**Discussion by MauiKano** [pschatzmann/arduino-audio-tools#755] [Q&A] [answered]
Hello,
I'd like to digitize two low quality audio streams via I2S and internal adc and analyze the two stream via fft. Does not have to run in parallel, e.g. only one stream is active in a given moment. The two audio sources are connected to GPIO34 and GPIO36. My code runs when I only configure one of them . With both configured I struggle with starting/stoping the streams and fft. Executing a fft.end() leads to a crash on the ESP32. How do I do that correctly ?

void setupFFT(int sr) {
  auto tcfg            = fft.defaultConfig();
  tcfg.length          = 2048;
  tcfg.channels        = 1; //2
  tcfg.sample_rate     = sr;  // a2dp_sink.sample_rate();;
  tcfg.bits_per_sample = 16;
  tcfg.callback        = &fftResult;
  tcfg.window_function = new BufferedWindow(new Hamming());
  fft.begin(tcfg);
}

void StartFFT(int sr) {
 setupFFT(sr);
}

void StopFFT() {
  fft.end();
}

void setupGeneralAdc(int pin) {
  auto cfg = Generaladc.defaultConfig(RX_MODE);
  cfg.sample_rate = 8000;
  cfg.channels = 1;
  cfg.setInputPin1(pin);
  Generaladc.begin(cfg);
}

void StartAdc() {
    StartFFT(8000);
    setupGeneralAdc(MIC_IN);
}

void StopAdc() {
  Generaladc.end();
  StopFFT();
}

**Comment by pschatzmann** [comment]
You would need to figure out why it is crashing: analyse the stacktrace !
You did not mention which of the many FFT implementation you were selecting...

Are you sure that you do not write any data to the closed fft ? 
That would explain your crashes

**Comment by MauiKano** [comment]
Analyzing the crashdump exeeds my capabilities. 
I use AudioRealFFT

It would already be a help when you could confirm how to properly start/stop AudioStreams and fft. I’m not sure if the ..end() method is the right way to do. E.g. if ..end() destroys all objects and allows me to create new ones during runtime

**Reply by pschatzmann** [reply, marked as answer]
I committed a correction that might help.
I did not reset the pointers to null so when the FFT object was created after calling end the necessary arrays and objects were not allocated any more...

**Reply by pschatzmann** [reply]
https://github.com/me-no-dev/EspExceptionDecoder

**Reply by MauiKano** [reply]
The Exception decoder can not decode my dump. I'm sure the mistake is on my side, I use this the first time. 
But still: Can I use fft.begin(cft) ... fft.end() .... fft.begin(cfg2)..... multiple times during runtime ?

hange Mode to: 4
starting FFT. Time per FFT: 46 Time per BPM: 920
Change Mode to: 5
starting FFT. Time per FFT: 256 Time per BPM: 5120
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x400d59ec  PS      : 0x00060b30  A0      : 0x800d5f1e  A1      : 0x3ffcb290  
A2      : 0x3ffdf448  A3      : 0x3ffee248  A4      : 0x3ffec238  A5      : 0x3ffbf35c  
A6      : 0x3ffc473c  A7      : 0x00006096  A8      : 0x000003ff  A9      : 0x00000004  
A10     : 0x3ffcb490  A11     : 0x80000001  A12     : 0x800815a4  A13     : 0x00000000  
A14     : 0x44ffe000  A15     : 0x00060c23  SAR     : 0x0000000a  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x000003ff  LBEG    : 0x40090950  LEND    : 0x4009095b  LCOUNT  : 0x00000000  


Backtrace: 0x400d59e9:0x3ffcb290 0x400d5f1b:0x3ffcb2b0 0x400d5fca:0x3ffcb2d0 0x400d6047:0x3ffcb2f0 0x400d51a2:0x3ffcb310 0x400d545f:0x3ffcb350 0x400d48bb:0x3ffcb380 0x400d4994:0x3ffcb3a0 0x400d6a8d:0x3ffcb3d0 0x400da5d9:0x3ffcb3f0




ELF file SHA256: 24811b65e718b4d9

Rebooting...

**Reply by pschatzmann** [reply]
Did you update the library? 
end() is releasing all resources and with the new version you can allocate them with begin() again...

I can't help you with the decoding because the decoder needs the compilation result to do the analysis.

**Reply by MauiKano** [reply]
I would think so. I have 2.0.7 for esp32 and downloaded the latest versions for audio-tools, RealFFT and A2DP from your github.

In the meantime I figured out how to decode the crash-dump. I did miss installing ..libcurses..

C: 0x400d59e4: ffft::FFTReal ::compute_direct_pass_1_2(float*, float const*) const at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/FFT/FFTReal.h line 1067
EXCVADDR: 0x000003ff

Decoding stack results
0x400d59e1: ffft::FFTReal ::compute_direct_pass_1_2(float*, float const*) const at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/FFT/FFTReal.h line 1067
0x400d5f13: ffft::FFTReal ::compute_fft_general(float*, float const*) const at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/FFT/FFTReal.h line 980
0x400d5fc2: ffft::FFTReal ::do_fft(float*, float const*) const at  .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/FFT/FFTReal.h line 980
0x400d603f: audio_tools::FFTDriverRealFFT::fft() at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/AudioRealFFT.h line 43
0x400d519a: audio_tools::AudioFFTBase::processSamples (void const*, unsigned int) at.../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/AudioFFT.h line 329
0x400d5457: audio_tools::AudioFFTBase::write(unsigned char const*, unsigned int) at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/AudioFFT.h line 147
0x400d48b3: audio_tools::StreamCopyT ::write(unsigned int, unsigned int&) at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioTools/AudioCopy.h line 273
0x400d498c: audio_tools::StreamCopy::copy (audio_tools::BaseConverter &) at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioTools/AudioCopy.h line 295
0x400d6a99: loop() at .../Projects/sketchbook.lx/rythmFCK/rythmFCK.ino line 1116
0x400da5e5: loopTask(void*) at .../.arduino15/packages/esp32/hardware/esp32/2.0.7/cores/esp32/main.cpp line 50

**Comment by pschatzmann** [comment]
I was just running a test and for me this is working w/o any issues. Are you sure that you installed the latest version when I told you that I have committed a correction ?

Here is my test case:
```C++

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioRealFFT.h" // or AudioKissFFT

AudioKitStream kit;  // Audio source
AudioRealFFT fft; // or AudioKissFFT
StreamCopy copier(fft, kit);  // copy mic to tfl
int channels = 2;
int samples_per_second = 44100;
int bits_per_sample = 16;
float value=0;

// display fft result
void fftResult(AudioFFTBase &fft){
    float diff;
    auto result = fft.result();
    if (result.magnitude>100){
        Serial.print(result.frequency);
        Serial.print(" ");
        Serial.print(result.magnitude);  
        Serial.print(" => ");
        Serial.print(result.frequencyAsNote(diff));
        Serial.print( " diff: ");
        Serial.println(diff);
    }
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup Audiokit
  auto cfg = kit.defaultConfig(RX_MODE);
  cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
  cfg.channels = channels;
  cfg.sample_rate = samples_per_second;
  cfg.bits_per_sample = bits_per_sample;
  kit.begin(cfg);

  // Setup FFT
  auto tcfg = fft.defaultConfig();
  tcfg.length = 8192;
  tcfg.channels = channels;
  tcfg.sample_rate = samples_per_second;
  tcfg.bits_per_sample = bits_per_sample;
  tcfg.callback = &fftResult;
  fft.begin(tcfg);
  copier.copyN(100);
  fft.end();

  fft.begin(tcfg);

}


void loop() { 
  copier.copy();
}
```

**Reply by MauiKano** [reply]
Mea Culpa, one one hand: I did not realize that you just recently updated the audio library. I did update it this morning and my sketch runs much more stable. E.g. I can switch between vaious fft configs (only the sampling rate changes). I can issue fft.begin(config1) .. fft.end().. fft.begin(config2).. fft.end() etc.Big step forward. Thanks for looking into this. BUT. First of all, my work-around with the older library was to just skip the fft.end() part. That seemed wrong but it made my sketch running perfectly fine. Now, after updating the library I still can provoke crashes. It happens after I stoped FFT with fft.end() and wanted to start a new one (setupFFT(scan rate) which is basically the same code as your FFT setup). The stack trace looks like listed below. Please note also that your example does not reflect my use-case. I have three input channels (two internal ADC, one bluetooth A2DP). Each has different sampling rate so when I switch between the input channels, I stop fft and start it again with the new config. Consequently, I have three copier objects that I switch between.


Decoding stack results
0x40083cfd: panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c line 408
0x4009313d: esp_system_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c line 137
0x40098e61: abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/abort.c line 46
0x40182ed3: __cxxabiv1::__terminate(void (*)()) at /builds/idf/crosstool-NG/.build/xtensa-esp32-elf/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc line 47
0x40182f1a: std::terminate() at /builds/idf/crosstool-NG/.build/xtensa-esp32-elf/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc line 57
0x4018325f: __cxxabiv1::__cxa_throw(void*, std::type_info*, void (*)(void*)) at /builds/idf/crosstool-NG/.build/xtensa-esp32-elf/src/gcc/libstdc++-v3/libsupc++/eh_throw.cc line 95
0x401831be: operator new(unsigned int) at /builds/idf/crosstool-NG/.build/xtensa-esp32-elf/build/build-cc-gcc-final/xtensa-esp32-elf/no-rtti/libstdc++-v3/include/bits/exception.h line 63
0x401832c9: operator new[](unsigned int) at /builds/idf/crosstool-NG/.build/xtensa-esp32-elf/src/gcc/libstdc++-v3/libsupc++/new_opv.cc line 32
0x400d5e39: audio_tools::FFTDriverRealFFT::begin(int) at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/AudioRealFFT.h line 26
0x400d4c80: audio_tools::AudioFFTBase::begin(audio_tools::AudioFFTConfig) at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/AudioFFT.h line 111
0x400d4d89: setupFFT(int) at .../Projects/sketchbook.lx/libraries/arduino-audio-tools/src/AudioLibs/AudioFFT.h line 43
0x400d4dea: StartFFT(int) at .../Projects/sketchbook.lx/rythmFCK/rythmFCK.ino line 737
0x400d4dff: StartMic() at .../Projects/sketchbook.lx/rythmFCK/rythmFCK.ino line 766
0x400d6b7a: ChangeMode(int) at .../Projects/sketchbook.lx/rythmFCK/rythmFCK.ino line 937
0x400d6bc5: rotary_onButtonClick(int) at .../Projects/sketchbook.lx/rythmFCK/rythmFCK.ino line 308
0x400d6ccf: rotary_loop() at .../Projects/sketchbook.lx/rythmFCK/rythmFCK.ino line 352
0x400d6cf9: loop() at .../Projects/sketchbook.lx/rythmFCK/rythmFCK.ino line 1155
0x400da86d: loopTask(void*) at .../.arduino15/packages/esp32/hardware/esp32/2.0.7/cores/esp32/main.cpp line 50

**Reply by pschatzmann** [reply]
You are just running out of RAM! You could resample to keep the sampling rate consistent.
Both A2DP and FFT are using up a lot of RAM, so be careful!

I also think that the different FFT implementations have different RAM requirements, so you could try to compare them in this regard. Or you could use a ESP32 which provides PSRAM.

**Reply by MauiKano** [reply]
Good point. I thought of that as well and I was not sure how that is handled. Obviously there is no memory management. But what happens during run-time when I start/stop FFT ? Is at start (fft.begin()) just memory allocated on the heap which will be freeed on fft.end() ?If this is the case, my sequence of fft.begin()...fft.end() should not consume condiderabely more RAM than just one call of fft.begin(). Anyway, having just one fft.begin() and one constante sample rate is defenitely doable for me and might even be the cleaner code.

**Comment by pschatzmann** [comment]
Yes, start() allocates memory both for the input & result array on the heap and calls the fft specific initialisation and end() is releasing it again.

**Reply by MauiKano** [reply]
hmm, well. Speaking of resampling. Did I miss anything basic ? I can not compile even the simple resample examples you provide. I get

      'audio_tools::ResampleStream' is not a template

at compile time when declaring something like

ResampleStream<int16_t> resample(sound);
CsvStream<int16_t> out(Serial); 
StreamCopy copier(out, resample);

**Comment by pschatzmann** [comment]
There were some changes and I might not have updated the examples:
ResampleStream does not need a template datatype any more.
You could also use the Formatconverter as decribed [here](https://github.com/pschatzmann/arduino-audio-tools/wiki/Converting-the-Data-Format), which was the reason for this change...
